GopherCAN.c

```diff
@@ -558,39 +563,32 @@ S8 mod_custom_can_func_state(GCAN_COMMAND_ID command_id, U8 state)
 //  designed to be called at high priority on 1ms loop
 void service_can_tx_hardware(CAN_HandleTypeDef* hcan)
 {
-	CAN_TxHeaderTypeDef tx_header;
 	CAN_MSG* message;
 	CAN_MSG_RING_BUFFER* buffer;
+    U32 tx_mailbox_num;
 
 	// With multiple busses, choose the correct bus buffer to be working with
-#ifdef MULTI_BUS
 	buffer = choose_tx_buffer_from_hcan(hcan);
-#else
-	buffer = &tx_buffer;
-#endif // MULTI_BUS
+
+    if (buffer->mutex != NULL) {
+        if(osMutexAcquire(buffer->mutex, MUTEX_TIMEOUT)) return;
+    }
+#if TARGET == F7XX || TARGET == F4XX
+    // disable TX interrupt to protect buffer
+    HAL_CAN_DeactivateNotification(hcan, CAN_IT_TX_MAILBOX_EMPTY);
```

Comment:
Mutex should mean we dont need to disable notifications

Reply:
This might be a bit tricky to handle because this function is called by both the TX complete interrupt and by tasks in software. There are different functions to get the mutex in both cases, so you might need to pass in if this function was called from ISR or task

Reply:
Related to my previous reply. I'm not sure how mutex ownership works between tasks and interrupts.
First two replies better explain the problem I'm thinking about:
https://forums.freertos.org/t/how-does-a-mutex-work-when-an-isr-using-the-protected-variable-triggers/12359/2